src/vivarium_census_prl_synth_pop/results_processing/formatter.py

```diff
@@ -162,3 +162,105 @@ def format_data_for_mapping(
+
+    # Get wide format of dependents - metadata for each guardian's dependents
+    dependents_wide = flatten_data(
+        data=df_dependents,
+        index_cols=["guardian_id", "tax_year"],
+        rank_col="simulant_id",
+        value_cols=[
+            "simulant_id",
+            "first_name",
+            "last_name",
+            "ssn_itin",
+            "copy_ssn",
+        ],
+    )
+    # Rename tax_dependents columns
+    dependents_wide = dependents_wide.add_prefix("dependent_").reset_index()
+    # Make sure we have all dependent columns if data does not have a guardian with 4 dependents
+    for i in range(2, 5):
+        if f"dependent_{i}_first_name" not in dependents_wide.columns:
+            dependents_cols = ["first_name", "last_name", "ssn_itin", "copy_ssn"]
+            for column in dependents_cols:
+                dependents_wide[f"dependent_{i}_{column}"] = np.nan
+
+    # Widen 1040 data (make one row for spouses that are joint filing)
+    df_joint_1040 = combine_joint_filers(df_1040)
+    # Merge tax dependents onto their guardians - we must do it twice, merge onto each spouse if joint filing
+    tax_1040_w_dependents = df_joint_1040.merge(
+        dependents_wide,
+        how="left",
+        left_on=["simulant_id", "tax_year"],
+        right_on=["guardian_id", "tax_year"],
+    )
+    # TODO: uncomment with mic-4244. Handle columns with dependents for both guardians
+    # tax_1040_w_dependents = tax_1040_w_dependents.merge(
+    #   dependents_wide, how="left",
+    #   left_on=["COLUMNS.spouse_simulant_id.name", "COLUMNS.tax_year.name"],
+    #   right_on=["COLUMNS.guardian_id.name", "COLUMNS.tax_year.name"])
+    # Rename "ssn_itin" columns to "ssn"
+    tax_1040_w_dependents.columns = tax_1040_w_dependents.columns.str.replace(
+        "ssn_itin", "ssn"
+    )
+
+    return tax_1040_w_dependents
+
+
+def flatten_data(
+    data: pd.DataFrame,
+    index_cols: str,
+    rank_col: str,
+    value_cols: List[str],
+    ascending: bool = False,
+) -> pd.DataFrame:
+    # Function that takes a dataset and widens (pivots) it to capture multiple metadata columns
+    # Example: simulant_id, dependdent_1, dependent_2, dependent_1_name, dependent_2_name, etc...
+    data = data.copy()
+    # fixme: find a better solution than the following call since applying lambda functions is slow
+    data["rank"] = (
+        data.groupby(index_cols, group_keys=False)[rank_col]
+        .apply(lambda x: x.rank(method="first", ascending=ascending))
+        .astype(int)
+    )
+    # TODO: Improve via mic-4244 for random sampling of dependents
+    # Choose 4 dependents
+    data = data.loc[data["rank"] < 5]
+    data["rank"] = data["rank"].astype(str)
+    flat = data.pivot(columns="rank", index=index_cols, values=value_cols)
+    flat.columns = ["_".join([pair[1], pair[0]]) for pair in flat.columns]
+
+    return flat
+
+
+def combine_joint_filers(data: pd.DataFrame) -> pd.DataFrame:
```

Comment:
I assume a followup PR will be removing all this from pseudopeople?

Reply:
Yes